lib/xrpc.rb

```diff
@@ -1,55 +1,130 @@
-# typed: false
 require "uri"
 require "httparty"
 require "json"
 
 module XRPC
-  def request(pds, endpoint_location, params)
-    Endpoint.new(pds, endpoint_location).get(params)
-  end
+  class Service
+    def initialize(pds)
+      @pds = pds
+    end
+
+    def call(endpoint, params = {}, body = nil)
+      endpoint_uri = URI("#{@pds}/xrpc/#{endpoint}?#{params.to_query}")
+      headers = { 'Content-Type': "application/json" }
 
-  module_function :request
+      response = if body
+          HTTParty.post(endpoint_uri, body: body.to_json, headers: headers)
+        else
+          HTTParty.get(endpoint_uri)
+        end
+
+      { encoding: response.headers["Content-Type"], body: JSON.parse(response.body) }
+    end
+  end
+end
 
+module XRPC
   class Endpoint
-    attr_reader :request_uri
+    attr_reader :id, :lexicon
 
-    def initialize(pds, endpoint_location, authenticated: false, token: nil)
-      @pds = pds
-      @endpoint_location = endpoint_location
-      @authenticated = authenticated
-      @headers = default_headers()
-      if token # Ideally, you shouldn't pass the token when creating the endpoint
-        @headers = default_authenticated_headers(token)
-      end
+    def initialize(id, lexicon)
+      @id = id
+      @lexicon = lexicon
     end
 
-    def default_headers
-      { "Content-Type" => "application/json" }
+    def call(service, params = {}, body = nil)
+      service.call(@id, params, body)
     end
+  end
+end
 
-    def authenticated?() @authenticated end
+module XRPC
+  class Lexicon
+    def initialize(lexicon)
+      @lexicon = lexicon
+    end
 
-    def default_authenticated_headers(access_token)
-      default_headers.merge({
-        Authorization: "Bearer #{access_token}",
-      })
+    def endpoint(id)
+      Endpoint.new(id, @lexicon)
     end
+  end
+end
 
-    def authenticate(token) # This is the proper place to authenticate with a token
-      # This is still a pretty weird way to authenticate, but it works (for now)
-      if not @authenticated == true
-        raise Error, "Non-authenticated endpoint cannot be authenticated"
-      end
-      @headers = default_authenticated_headers(token)
+module XRPC
+  class Registry
```

Comment:
"Registry"? I'm not sure quite what I was thinking with that name, I don't know if it's the best name for this or a bad one, but we'll see. It might be best to keep it, actually. I'm not sure 🤔 

Thinking out loud, lol